feat: Switch from Yoga v2 to Apollo Server v4

Moumouls opened the pull request:
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

## Issue
<!-- Add the link to the issue that this PR closes. -->

Closes: #8728

## Approach
<!-- Describe the changes in this PR. -->
- Install graphql upload to support upload
- Install GraphIQL from Yoga team since GraphIQL is standard, and the yoga team graphiql is fine
- Take advantage of X-Parse-Application-Id as CSRF protection [https://www.apollographql.com/docs/router/configuration/csrf/](https://www.apollographql.com/docs/router/configuration/csrf/)
- apollo/server is more focused on NodeJS and have dedicated expressHandler, i hope it will be more stable in complex envs.
- Now we have end to end stream system from client to bucket/storage system through GQL Upload API
- Disable the cache control since it's not used at the moment and can slightly help for performance
- Remove all it_only_node < 17, now upload works on major NodeJS version without issues
- Added mime type package to auto add extension to filename without extension since it's the current behavior, without this we will have a breaking change

Performance tests:
- Checked performance on ParseServerGraphQL test suite and complete test suite in my company with a lot of graphql requests, performance are close and quite the same.

## Tasks
<!-- Delete tasks that don't apply. -->

- [x] Add tests
- [x] Add changes to documentation (guides, repository pages, code comments)
- [x] Add [security check](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#security-checks)
- [x] Add new Parse Error codes to Parse JS SDK <!-- no hard-coded error codes in Parse Server -->

Moumouls committed:
feat: switch to apollo and stream file upload

Moumouls committed:
test: max file size

mtrezza commented:
I notice that this also requires Node 16, so we'll wait with merging info alpha until Nov.

mtrezza reviewed (commented):
Apart from the requirement of Node >= 16, does this have any breaking changes? It's fine if it does, since we can only release it with Parse Server 7 (2024) anyway.

Moumouls committed:
feat: add auto extension

Moumouls commented:
@mtrezza no breaking changen may be the error message of graphql upload, but developers should not rely on message.

Also i added mime type detection, since it's currently supported.

Yes @mtrezza it's not a problem to keep this PR for 2024 :)

mtrezza commented on package.json:
```diff
@@ -48,6 +48,7 @@
     "lodash": "4.17.21",
     "lru-cache": "9.1.1",
     "mime": "3.0.0",
+    "mime-types": "2.1.35",
```
We already have [mime](https://www.npmjs.com/package/mime), it seems that it's similar to [mime-types](https://www.npmjs.com/package/mime-types). I think if possible we should decide for one of the two if they provide the same functionality.

mtrezza commented:
Could you suggest an entry for the changelog?

The current title would only be a refactor, but from the PR comment it seems to include many more improvements, so it would be good to add them to the changelog. You don't have to write that into the PR title, I'll add the long version in the commit message and we keep the PR title short.

mtrezza committed:
Merge branch 'alpha' into moumouls/switch-to-apollo-and-file-stream

mtrezza commented:
@Moumouls While this PR is still "recent" and to prepare it for merging in Nov, do you think you could address the mime dependency issue? Otherwise I'm afraid this PR will go stale and not get merged for a long time if no-one picks it up.

mtrezza committed:
Merge branch 'alpha' into moumouls/switch-to-apollo-and-file-stream

Moumouls commented:
Hi @mtrezza, we should definitely merge this one. It works really well.

What is the mime dep issue. How is it linked to this Pr ?

Just tell me when I need to refresh the branch, and I'll be happy to fix last details to get it merged !

mtrezza commented:
See https://github.com/parse-community/parse-server/pull/8727#discussion_r1313639096.

Moumouls commented:
Nice catch @mtrezza , I'll take a look

Moumouls committed:
feat: fix node 18 and move to apollo sandbox

Moumouls committed:
Merge branch 'alpha' into moumouls/switch-to-apollo-and-file-stream

Signed-off-by: Antoine Cormouls <contact.antoine.cormouls@gmail.com>

Moumouls commented:
Hi @mtrezza i replaced last yoga dep with the CDN version of Apollo Sandbox: https://www.apollographql.com/docs/graphos/explorer/sandbox 

Also fixed your feedback to use current mime dep

My team use this branch in production since now 2 months everything works well.

Moumouls committed:
test: update node 18

mtrezza commented on src/GraphQL/ParseGraphQLServer.js:
```diff
@@ -79,35 +93,55 @@ class ParseGraphQLServer {
-              'X-Parse-Master-Key': this.parseServer.config.masterKey,
-            }),
-          })
+          `<div id="sandbox" style="position:absolute;top:0;right:0;bottom:0;left:0"></div>
+          <script src="https://embeddable-sandbox.cdn.apollographql.com/_latest/embeddable-sandbox.umd.production.min.js"></script>
+          <script>
+           new window.EmbeddedSandbox({
+             target: "#sandbox",
+             endpointIsEditable: false,
+             initialEndpoint: "${JSON.stringify(this.config.graphQLPath)}",
+             handleRequest: (endpointUrl, options) => {
+              return fetch(endpointUrl, {
+                ...options,
+                headers: {
+                    ...options.headers,
+                    'X-Parse-Application-Id': "${JSON.stringify(this.parseServer.config.appId)}",
+                    'X-Parse-Master-Key': "${JSON.stringify(this.parseServer.config.masterKey)}",
+                },
+              })
+            },
+           });
+           // advanced options: https://www.apollographql.com/docs/studio/explorer/sandbox#embedding-sandbox
+          </script>`
```
Is this safe?

Moumouls replied:
it's safe @mtrezza , also the playground should be only used in dev, never in production, it's mentionned in the docs

mtrezza commented on src/GraphQL/loaders/filesMutations.js:
```diff
@@ -1,33 +1,63 @@
 import * as defaultGraphQLTypes from './defaultGraphQLTypes';
 import logger from '../../logger';
 
+// Handle graphql file upload and proxy the file upload
+// to the graphql server url specified in the config
+// We do not call directly createFile from the Parse Server
+// to leverage the standard file upload mechanism
```
```suggestion
// Handle GraphQL file upload and proxy file upload to GraphQL server url specified in config;
// `createFile` is not directly called by Parse Server to leverage standard file upload mechanism
```
```suggestion
// Handle graphql file upload and proxy the file upload
// to the graphql server url specified in the config
// We do not call directly createFile from the Parse Server
// to leverage the standard file upload mechanism
```

kekami commented on src/GraphQL/ParseGraphQLServer.js:
```diff
@@ -57,8 +58,21 @@ class ParseGraphQLServer {
+        // needed since we use graphql upload
+        requestHeaders: ['X-Parse-Application-Id'],
+      },
+      introspection: true,
```
Introspection should probably be disabled by default for security reasons. Perhaps a config variable.

Moumouls replied:
Hi @kekami , you are right but it needs a dedicated PR, i created the FR: https://github.com/parse-community/parse-server/issues/8800

Moumouls committed:
Update src/GraphQL/loaders/filesMutations.js

Co-authored-by: Manuel <5673677+mtrezza@users.noreply.github.com>
Signed-off-by: Antoine Cormouls <contact.antoine.cormouls@gmail.com>

mtrezza commented:
@Moumouls We have begun merging the breaking change PRs for Parse Server 7. Could you please resolve the conflict in this PR so we can merge it?

mtrezza commented:
Closing via https://github.com/parse-community/parse-server/pull/8959
